test/e2e/user_workload_monitoring_test.go

```diff
@@ -633,6 +641,134 @@ func assertTenancyForMetrics(t *testing.T) {
 	}
 }
 
+// assertTenancyForRules ensures that a tenant can access rules from her namespace (and only from this one).
+func assertTenancyForRules(t *testing.T) {
+	const testAccount = "test-rules"
+
+	_, err := f.CreateServiceAccount(userWorkloadTestNs, testAccount)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	// Grant enough permissions to the account so it can read rules.
+	_, err = f.CreateRoleBindingFromClusterRole(userWorkloadTestNs, testAccount, "monitoring-rules-view")
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	var token string
+	err = framework.Poll(5*time.Second, 5*time.Minute, func() error {
+		token, err = f.GetServiceAccountToken(userWorkloadTestNs, testAccount)
+		if err != nil {
+			return err
+		}
+		return nil
+	})
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	// The tenancy port (9093) is only exposed in-cluster so we need to use
+	// port forwarding to access kube-rbac-proxy.
+	host, cleanUp, err := f.ForwardPort(t, "thanos-querier", 9093)
+	if err != nil {
+		t.Fatal(err)
+	}
+	defer cleanUp()
+
+	client := framework.NewPrometheusClient(
+		host,
+		token,
+		&framework.QueryParameterInjector{
+			Name:  "namespace",
+			Value: userWorkloadTestNs,
+		},
+	)
+
+	t.Logf("Retrieving rules")
```

Comment:
nit: test logs will not be emitted at the time of invocation, but after the test, hence imho this could be removed, but can be done as a follow-up too.

Reply:
Most probably a left-over piece. I'll remove it.